### ISSUE TITLE
Hercules crash during QETH HSCH processing

### ISSUE THREAD

**Issue by Juergen-Git** [SDL-Hercules-390/hyperion#690]
___
_**Note:** this issue is closely related to issue [\#723: "Qeth: at shutdown "buffer overflow detected""](https://github.com/SDL-Hercules-390/hyperion/issues/723)._
___
Fish, as promised, here the issue concerning the QETH problems we recently discussed:

The Hercules QETH device causes Hercules to crash, when the operating system issues a HSCH against the data device while it is still active/busy.

That behavior  typically occurs on Windows systems, when the QETH device is using SoftDevLabs' CTCI-WIN product, and when the LSO functionality of the host system's Ethernet adapter isn't handled correctly by CTCI-WIN.

While there exists a workaround for that specific situation (simply disabling LSO on the host's adapter), the resulting Hercules crash should be seen as an independent problem. It can be expected that issuing `HSCH` (Halt Subchannel) against the QETH data device in the "wrong moment" will generally trigger the problem.

Analysis shows that the crash gets triggered when `qeth_halt_data_device` gets called and doesn't return because the device doesn't honor the `QDSIG_HALT` signal.

That consequently leads to a deadlock (as `qeth_halt_data_device` holds both `INTLOCK `and `dev->lock`, thus preventing some other CPU thread from being able to acquire `INTLOCK` for its own purposes) which eventually leads to a forced crash the moment our `watchdog_thread` system monitoring thread  eventually detects the deadlock/hang.

The deadlock _(and thus the crash)_ can be avoided by the following quick and dirty patch, which is only meant to help further analysis. As it doesn't solve the underlying problem, it is not intended for production use.

```diff
From: Juergen Winkelmann <juergen.winkelmann@pebble-beach.ch>
Date: Sat, 9 Nov 2024 13:48:25 +0100
Subject: [PATCH] temporary solution to prevent Hercules crash during QETH HSCH processing

---
 qeth.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/qeth.c b/qeth.c
index c345f99c..be06a4fa 100644
--- a/qeth.c
+++ b/qeth.c
@@ -3812,11 +3812,19 @@ static void qeth_halt_data_device( DEVBLK* dev, OSA_GRP* grp )
 
             DBGTRC( dev, "Halting data device" );
             {
+                struct timespec waittime;
+                struct timeval  now;
+                gettimeofday( &now, NULL );
+                waittime.tv_sec  = now.tv_sec  + 5;
+                waittime.tv_nsec = now.tv_usec * 1000;
+
                 /* Ask, then wait for, the Activate Queues loop to exit */
                 PTT_QETH_TRACE( "b4 halt data", 0,0,0 );
                 VERIFY( qeth_write_pipe( grp->ppfd[1], &sig ) == 1);
-                wait_condition( &grp->qdcond, &grp->qlock );
-                dev->scsw.flag2 &= ~SCSW2_Q;
+                if (!timed_wait_condition( &grp->qdcond, &grp->qlock, &waittime ))
+                    dev->scsw.flag2 &= ~SCSW2_Q;
+                else
+                    logmsg("QETH Timeout during halt processing: SCSW2_Q NOT cleared\n");
                 PTT_QETH_TRACE( "af halt data", 0,0,0 );
             }
             DBGTRC( dev, "Data device halted" );
-- 
2.34.1

```

Cheers
Jürgen

**Comment by Fish-Git** [comment]
LBO? &nbsp;I've never heard of that. &nbsp;Do you maybe mean [**LSO**](https://en.wikipedia.org/wiki/TCP_offload_engine#Large_send_offload)? &nbsp;(Large **Segment** Offloading?)

_(I know! I know! It's hard to keep track of all these abbreviations!)_

**Comment by Juergen-Git** [comment]
Right you are… I didn’t bother to look it up again and large block offload (LBO) was, what dropped out of my memory. Getting old 🙁…

**Comment by Juergen-Git** [comment]
I'm not sure it's a Windows only problem: Yes, I've observed it so far under Windows only, in conjunction with CTCI-WIN. However, with the current knowledge I'd not bet that this cannot happen in other environments too 😉. Probably, it doesn't need more than just feeding nonsense into the device until the IP stack gives up and tries to bail out using HSCH...

**Comment by Fish-Git** [comment]
You're probably right. I've updated the text of the 'W' GitHub issue label. I promise to get to this issue as soon as I can but still desperately trying to get caught up with Hercules and serious personal issues backlog, _(My wife had another serious heart attack!)_

**Comment by Juergen-Git** [comment]
As I said earlier: This issue has absolutely no priority. Health, your wife's and your's is most important! Wishing your wife all the best, take care!

**Comment by wrljet** [comment]
I'm very sorry to hear that, Fish.  I wish the best to you both.

Bill

**Comment by Juergen-Git** [comment]
Hi Fish

thanks for the analysis!
>
> Given my own analysis, I doubt your patch will accomplish much of anything. (Have you tried it? Does it work?)
>
The only thing my patch accomplishes is to prevent the deadlock. That means, after applying the patch Hercules will no longer crash when the problem occurs. The problem itself doesn't get fixed by the patch, and thus the device ends up being no longer usable once the problem occured. Even detaching and attaching it from/to Hercules doesn't bring it back to life. And yes, I've tried it and it works exactly to that point. My idea was, to allow further analysis when Hercules is still running -- but I'm not exactly an in depth windows debugger ;-). And given, that the problem doesn't seem to occur widely across the Windows/QETH community I still think it has verly low priority only.

Cheers
Jürgen

**Comment by Fish-Git** [comment]
> The only thing my patch accomplishes is to prevent the deadlock.

Wrong! And _that_ was my point! It _doesn't_ prevent the deadlock! &nbsp;_(and will **never** do so! -- by itself)_

The reason it doesn't prevent the deadlock (by itself) is because your patch is _still calling_ the `qeth_write_pipe` function, and it is _that_ function that is _never returning_ (presumably). So your `timed_wait_condition` will never be executed! So it _doesn't_ prevent the deadlock at all. By itself, it essentially accomplishes absolutely nothing.

Now, once I eventually fix the loop in the `qeth_write_pipe` function so that it doesn't loop forever and instead eventually returns, _**THEN**_ your `timed_wait_condition` will indeed prevent the deadlock from occurring.

But again, _by itself_, it accomplishes nothing. It does not, and will not, fix the deadlock by itself. The deadlock can _only_ be fixed by applying _BOTH_ proposed fixes: my fix to prevent `qeth_write_pipe` from looping forever, and your `timed_wait_condition` fix to not wait forever for the signal acknowledging that the HSC request has completed. Only when _both_ fixes are in place will the deadlock be prevented. If only one fix or the other is applied, the deadlock will _still occur!_ _**Both**_ fixes are needed to prevent the deadlock.

Does that make sense now?

> the problem doesn't seem to occur widely across the Windows/QETH community I still think it has very low priority only.

Agreed 100%. &nbsp;But you know me: &nbsp;I hate _any_ bug in Hercules, no matter how minor or insignificant it may be.

Especially any bug related to networking, as such bugs tend to make my CTCI-WIN product look bad. &nbsp;`;-)`

p.s. I'm currently working on _trying_ to make changes to my CTCI-WIN product to also report an adapter's "LSO" (Large Segment Offload) setting, and issue a warning should it appear that the adapter has it enabled. This, along with my existing checksum offloading reporting/handling,  _should_ (I hope!) prevent such hard to diagnose problems from occurring, as well as significantly reduce my own product support workload. &nbsp;_(Most of the support questions I receive these days are specifically related to issues exactly like this: incorrect adapter offload settings.)_

**Comment by Juergen-Git** [comment]
Hi Fish

> The reason it doesn't prevent the deadlock (by itself) is because your patch is still calling the qeth_write_pipe function, and it is that function that is never returning (presumably). So your timed_wait_condition will never be executed! So it doesn't prevent the deadlock at all. By itself, it essentially accomplishes absolutely nothing.

Logic-wise I'm 100% with you. However, how to explain then this:

```diff
08:51:25 HHC03800I 0:0401 QETH: Adapter mode set to Layer 3
08:51:25 HHC04100I TunTap64.dll version 3.7.0.6070 initiated
08:51:25 HHC00901I 0:0401 QETH: Interface tun0, type TUN opened
08:51:26 HHC03997I 0:0401 QETH: tun0: using MAC address 02-00-5E-A8-4B-6F
08:51:26 HHC03997I 0:0401 QETH: tun0: using IP address 192.168.50.245
08:51:26 HHC03997I 0:0401 QETH: tun0: using subnet mask 255.255.255.0
08:51:26 HHC03997I 0:0401 QETH: tun0: using MTU 1500
08:51:26 HHC03997I 0:0401 QETH: tun0: using drive MAC address 02:90:31:cf:95:7a
08:51:26 HHC03997I 0:0401 QETH: tun0: using drive IP address fe80::290:31ff:fecf:957a
08:51:26 HHC03805I 0:0401 QETH: tun0: Register guest IP address 192.168.50.246
09:04:40 HHC01603I qproc
09:04:40 HHC17007I NumCPU = 02, NumVEC = 00, ReservedCPU = 00, MaxCPU = 02
09:04:40 HHC17008I Avgproc  02 004%; MIPS[   3.60]; SIOS[    10]
09:04:40 HHC17009I PROC CP00 - 009%; MIPS[   3.55]; SIOS[     8] - Host Kernel(00:00:01.797) User(00:00:57.406)
09:04:40 HHC17009I PROC IP01 - 000%; MIPS[   0.05]; SIOS[     2] - Host Kernel(00:00:01.188) User(00:01:33.984)
09:04:40 HHC17010I  - Started        : Stopping        * Stopped
09:04:52 HHC01603I ipending
09:04:52 HHC00850I Processor CP00: CPUint=00000001 (State:00000001)&(Mask:DF00EE2B)
09:04:52 HHC00851I Processor CP00: interrupt not pending
09:04:52 HHC00852I Processor CP00: I/O interrupt not pending
09:04:52 HHC00853I Processor CP00: clock comparator not pending
09:04:52 HHC00854I Processor CP00: CPU timer not pending
09:04:52 HHC00855I Processor CP00: interval timer not pending
09:04:52 HHC00856I Processor CP00: ECPS vtimer not pending
09:04:52 HHC00857I Processor CP00: external call not pending
09:04:52 HHC00858I Processor CP00: emergency signal not pending
09:04:52 HHC00859I Processor CP00: machine check interrupt not pending
09:04:52 HHC00860I Processor CP00: service signal not pending
09:04:52 HHC00861I Processor CP00: mainlock held: no
09:04:52 HHC00862I Processor CP00: intlock held: no
09:04:52 HHC00863I Processor CP00: waiting for intlock: no
09:04:52 HHC00864I Processor CP00: lock not held
09:04:52 HHC00866I Processor CP00: state STARTED
09:04:52 HHC00867I Processor CP00: instcount 15999827536
09:04:52 HHC00868I Processor CP00: siocount 84684
09:04:52 HHC00869I Processor CP00: psw 0706000000000000 0000000000000000
09:04:52 HHC00850I Processor IP01: CPUint=80000001 (State:80000001)&(Mask:DF00E62B)
09:04:52 HHC00851I Processor IP01: interrupt pending
09:04:52 HHC00852I Processor IP01: I/O interrupt not pending
09:04:52 HHC00853I Processor IP01: clock comparator not pending
09:04:52 HHC00854I Processor IP01: CPU timer not pending
09:04:52 HHC00855I Processor IP01: interval timer not pending
09:04:52 HHC00856I Processor IP01: ECPS vtimer not pending
09:04:52 HHC00857I Processor IP01: external call not pending
09:04:52 HHC00858I Processor IP01: emergency signal not pending
09:04:52 HHC00859I Processor IP01: machine check interrupt not pending
09:04:52 HHC00860I Processor IP01: service signal not pending
09:04:52 HHC00861I Processor IP01: mainlock held: no
09:04:52 HHC00862I Processor IP01: intlock held: no
09:04:52 HHC00863I Processor IP01: waiting for intlock: no
09:04:52 HHC00864I Processor IP01: lock not held
09:04:52 HHC00866I Processor IP01: state STARTED
09:04:52 HHC00867I Processor IP01: instcount 40571800651
09:04:52 HHC00868I Processor IP01: siocount 24774
09:04:52 HHC00869I Processor IP01: psw 0706000000000000 0000000000000000
09:04:52 HHC00870I config mask 00000003 started mask 00000003 waiting mask 00000003
09:04:52 HHC00871I syncbc mask 00000000 
09:04:52 HHC00872I signaling facility not busy
09:04:52 HHC00873I TOD lock not held
09:04:52 HHC00874I mainlock not held; owner ffff
09:04:52 HHC00875I intlock not held; owner ffff
09:04:52 HHC00876I ioq lock not held
09:04:52 HHC00884I Channel Report queue: (empty)
09:04:52 HHC00880I device 0:0400: status busy local
09:04:52 HHC00880I device 0:0402: status busy local
09:04:52 HHC00880I device 0:0A82: status reserved local
09:04:52 HHC00881I I/O interrupt queue: (NULL)
09:04:57 HHC03997I 0:0401 QETH: tun0: not using MAC address 02-00-5E-A8-4B-6F
09:04:57 HHC03997I 0:0401 QETH: tun0: not using IP address 192.168.50.245
09:04:57 HHC03997I 0:0401 QETH: tun0: not using subnet mask 255.255.255.0
09:04:57 HHC03997I 0:0401 QETH: tun0: not using MTU 1500
- 09:05:02 QETH Timeout during halt processing: SCSW2_Q NOT cleared
09:05:16 HHC01603I ipending
09:05:16 HHC00850I Processor CP00: CPUint=00000001 (State:00000001)&(Mask:DF00EE2B)
09:05:16 HHC00851I Processor CP00: interrupt not pending
09:05:16 HHC00852I Processor CP00: I/O interrupt not pending
09:05:16 HHC00853I Processor CP00: clock comparator not pending
09:05:16 HHC00854I Processor CP00: CPU timer not pending
09:05:16 HHC00855I Processor CP00: interval timer not pending
09:05:16 HHC00856I Processor CP00: ECPS vtimer not pending
09:05:16 HHC00857I Processor CP00: external call not pending
09:05:16 HHC00858I Processor CP00: emergency signal not pending
09:05:16 HHC00859I Processor CP00: machine check interrupt not pending
09:05:16 HHC00860I Processor CP00: service signal not pending
09:05:16 HHC00861I Processor CP00: mainlock held: no
09:05:16 HHC00862I Processor CP00: intlock held: no
09:05:16 HHC00863I Processor CP00: waiting for intlock: no
09:05:16 HHC00864I Processor CP00: lock not held
09:05:16 HHC00866I Processor CP00: state STARTED
09:05:16 HHC00867I Processor CP00: instcount 16070442290
09:05:16 HHC00868I Processor CP00: siocount 84769
09:05:16 HHC00869I Processor CP00: psw 0706000000000000 0000000000000000
09:05:16 HHC00850I Processor IP01: CPUint=00000001 (State:00000001)&(Mask:DF00E62B)
09:05:16 HHC00851I Processor IP01: interrupt not pending
09:05:16 HHC00852I Processor IP01: I/O interrupt not pending
09:05:16 HHC00853I Processor IP01: clock comparator not pending
09:05:16 HHC00854I Processor IP01: CPU timer not pending
09:05:16 HHC00855I Processor IP01: interval timer not pending
09:05:16 HHC00856I Processor IP01: ECPS vtimer not pending
09:05:16 HHC00857I Processor IP01: external call not pending
09:05:16 HHC00858I Processor IP01: emergency signal not pending
09:05:16 HHC00859I Processor IP01: machine check interrupt not pending
09:05:16 HHC00860I Processor IP01: service signal not pending
09:05:16 HHC00861I Processor IP01: mainlock held: no
09:05:16 HHC00862I Processor IP01: intlock held: no
09:05:16 HHC00863I Processor IP01: waiting for intlock: no
09:05:16 HHC00864I Processor IP01: lock not held
09:05:16 HHC00866I Processor IP01: state STARTED
09:05:16 HHC00867I Processor IP01: instcount 40579796477
09:05:16 HHC00868I Processor IP01: siocount 24789
09:05:16 HHC00869I Processor IP01: psw 0706000000000000 0000000000000000
09:05:16 HHC00870I config mask 00000003 started mask 00000003 waiting mask 00000003
09:05:16 HHC00871I syncbc mask 00000000 
09:05:16 HHC00872I signaling facility not busy
09:05:16 HHC00873I TOD lock not held
09:05:16 HHC00874I mainlock not held; owner ffff
09:05:16 HHC00875I intlock not held; owner ffff
09:05:16 HHC00876I ioq lock not held
09:05:16 HHC00884I Channel Report queue: (empty)
09:05:16 HHC00880I device 0:0400: status busy local
09:05:16 HHC00881I I/O interrupt queue: (NULL)
```
The message in red above, issued by my patch once `timed_wait_condition` runs into its timeout, should never be issued if the theory is correct right that the `timed_wait_condition` never gets executed. So, what are we missing?

Cheers
Jürgen

**Comment by Juergen-Git** [comment]
> Which is why I'm curious about how you were able to reproduce the problem? If it can be reproduced, I'd very much like to try it for myself! How did you do it?

This is still the original problem: It occurs, reliably, when I open my slightly modified CICS web interface IVP website using HTTPS on a Windows system with LSO enabled on the network adapter. You can see how the result should look like when opening https://cicsweb:webcics@cpblcy3b.pebble-beach.ch:3005/CICS/CWBA/DFH$WB1A. This one runs on a Linux instance and it is exactly this interaction that produces the error on a Windows instance with LSO enabled.

So, how can you repoduce it for yourself? You have to setup a CICS instance with its web interface activated over HTTPS. To do so, you need a certificate (self signed suffices) and make this available to the CICS web server through a TLS policy via PAGENT. Once you have the web interface up and running, you can replace the original IVP website by my modified one, and then you will see the problem by yourself.

The point here is that this is a reliable way to reproduce the problem, but it most probably has nothing to do with the problem. As I said earlier, I'm convinced that there must be an easier way to reproduce it without having to set up a full blown CICS instance with web interface. One would somehow need to produce that same I/O pattern on the QETH interface, and I'd almost bet that the problem would then occur.

But nonetheless, let me know whether you want to go through the hassle of defining a certificate, creating a TLS rule and setting up the CICS web interface -- then I'll try to provide you off list with the information necessary on how to do this.

**Comment by Fish-Git** [comment]
> > Which is why I'm curious about how you were able to reproduce the problem? If it can be reproduced, I'd very much like to try it for myself! How did you do it?
> 
> This is still the original problem: It occurs, reliably, when I ...

Oh! I wasn't aware of that. Sorry. My bad. I thought this was a one-time occurrence. I wasn't aware that it could be _reliably_ reproduced at will. That's great! That makes things much easier.

> I'm convinced that there must be an easier way to reproduce it without having to set up a full blown CICS instance with web interface. One would somehow need to produce that same I/O pattern on the QETH interface, and I'd almost bet that the problem would then occur.

And therein is the rub: while it may be true that there is _probably_ another, _perhaps_ easier, way to reproduce the problem, we don't know that for certain, _AND_ discovering that different way (since we don't currently know what the other way is) would likely take many attempts to find it. (And then there's no guarantee that we would _ever_ find it. It's just a presumption after all.)

Therefore I would prefer using the same technique that we already know _does_ work, and works _reliably_. Why waste our time trying to discover a _new_ way to reproduce the problem, when we already know of a way that _reliably_ reproduces it each and every time?

> ... let me know whether you want to go through the hassle of defining a certificate, creating a TLS rule and setting up the CICS web interface -- then I'll try to provide you off list with the information necessary on how to do this.

I would like that very much! Yes!

While I'm sure we could work together on this bug with me providing you some research patches to fump information at certain points in the code, you then rebuilding Herc with my patches applied, running your repro and sending me the results (and then lathering, rinsing and repeating, as many times as necessary for me to get to the root of the problem and providing a fix for it (and then you testing it for me)), it would be much more convenient (and likely faster too) for both me and you for you to simply help me set things up on my own system (systems?) here so I can easily reproduce for myself whenever I need to.

Given that the problem involves both my CTCI-WIN product and having LSO mistakenly enabled on the adapter too (a common user error that usually involves a support call from some of my customers), it would be much better for me to be able to reproduce the bug for myself -- _at will_ -- too, as that would allow me to test some changes I've been wanting to _(trying to!)_ do for a long time now. If I can get my CTCI-WIN product to detect and warn my customers about their LSO settings, all the better for me too, and not just Hercules.

So let's do that: work together offline so I can obtain the needed help from you on setting up my own system so I can reproduce the problem myself. Once I can repro the problem on demand myself, I'll be able to find and fix the problem much more quickly and reliably than trying to debug this issue remotely. Agreed? Agreed!

So I'll wait for you to contact me off list. &nbsp;`:)`

Thanks, Jürgen. You're a nice guy.

**Comment by Fish-Git** [comment]
It look a while, but with Jürgen's help, I was able to setup my z/OS to recreate his problem.

I am only able to recreate it on my Windows 11 system however (which is a VMware virtual machine, currently running 21H2(?) I think it's called?), but I cannot recreate it on my Windows 7 host system.

So now that I am able to reproduce the crash at will _(i.e. whenever I want to)_, I will _hopefully_ be able to eventually track down the root cause and get the problem fixed.

Seeing how this is a Low Priority issue, I am going to change the GitHub status from "In Progress" to "OnGoing", since I suspect it's going to take me a while.

(Besides, there are other more important issues I could be working on, and if I try to work on too many things at once, I only end up making mistakes. So I'm putting this issue on hold for a while.)

**Comment by Fish-Git** [comment]
I haven't forgotten about you, Jürgen! &nbsp; I'm still working on your issue, albeit at a lower priority.

Hopefully, now that I have my needed Windows 11 virtual machines setup and am able to reproduce your issue on demand myself, it won't be long before I'll have a proposed fix for you to try. (*)

(*)&nbsp; _Famous last words, eh?_ &nbsp; `:-P`

**Comment by Juergen-Git** [comment]
no need to hurry, though 😇

**Comment by Juergen-Git** [comment]
Thanks, Fish! I'll go after this tomorrow 😊

**Comment by Juergen-Git** [comment]
Hi Fish

works as advertised 😊, thanks a lot!

Concerning the detach/attach cycle:

I'm seeing the following messages in the Hercules log:

18:12:36 HHC01603I detach 400
18:13:11 HHC01603I attach 0400.3 QETH hwaddr 02-00-5E-A8-4B-6F ipaddr 192.168.50.245 netmask 255.255.255.0 chpid F0
**18:13:11 HHC90000D DBG: ** Win32 porting error: invalid call to 'w32_select' from qeth.c(861): mixed set(s)
18:13:11 HHC90000D DBG: ** Win32 porting error: invalid call to 'w32_select' from qeth.c(861): mixed set(s)**
18:14:36 HHC00905I 0:0400 QETH: Halt or clear recognized for read device
18:14:36 HHC00905I 0:0400 QETH: Halt or clear completed for read device
18:14:36 HHC00905I 0:0402 QETH: Halt or clear recognized for data device
18:14:36 HHC00905I 0:0402 QETH: Halt or clear completed for data device
18:14:36 HHC03800I 0:0401 QETH: Adapter mode set to Layer 3
18:14:36 HHC00901I 0:0401 QETH: Interface **tun1**, type TUN opened
18:14:36 HHC03997I 0:0401 QETH: tun1: using MAC address 02-00-5E-A8-4B-6F
18:14:36 HHC03997I 0:0401 QETH: tun1: using IP address 192.168.50.245
18:14:36 HHC03997I 0:0401 QETH: tun1: using subnet mask 255.255.255.0
18:14:36 HHC03997I 0:0401 QETH: tun1: using MTU 1500
18:14:36 HHC03997I 0:0401 QETH: tun1: using drive MAC address 02:90:31:cf:95:7a
18:14:36 HHC03997I 0:0401 QETH: tun1: using drive IP address fe80::290:31ff:fecf:957a
18:14:36 HHC03805I 0:0401 QETH: tun1: Register guest IP address 192.168.50.246

At first sight, I can't see a problem in the select() call. Can it be that the read file descriptor set (rdset) gets stomped over somewhere during the cycle? Might this even be a hint to the problem that prevents automatic recovery?

Also notable is that the interface number gets incremented (it was tun0 originally but after the detach/attach cycle it becomes tun1. A hint too, or expected behavior?

Cheers
Jürgen

**Comment by Fish-Git** [comment]
Weird. That never happened to me that I noticed.

**Question:** did you disable LSO between the Hercules detach/attach? When I tested, I didn't bother to do that.(*) Maybe that's what causing it? Let me give it a try...

Okay, I did _not_ get any "** Win32 porting error" message, but yes, I _did_ notice that the "tun" value did indeed change from its original "tun0" value to "tun1".

Without bothering to dig into the code, I suspect that's just a normal side effect of re-attaching the OSA device again. From our TunTap logic's point of view, it's a brand new tuntap device (interface), so it makes sense that it would assign a brand new name to it. \<shrug\>

I can only speculate that the `HHC90000D ** Win32 porting error...` message(s) are just bad timing. Notice that they're issued immediately after the `attach` is done in Hercules, and well before the halt subchannel occurs. Without bothering to take the time to actually step through the code, I highly suspect it's just a simple matter of the original socket becoming invalid as a result of the original `detach`, thereby causing our `socket_is_socket` function to classify it as a non-socket file descriptor. That, along with the _other_ file descriptor in the set being checked still being a valid socket (i.e. the socket "pipe" used to request and acknowledge HSCH requests), is what caused the "mixed set" messages to be issued.

I suppose I could make another minor tweak to `qeth.c`s ACTIVATE QUEUES logic to _remove_ the OSA device's file descriptor (`grp->ttfd`) from the select set before calling `qeth_select` if the interface is not "up" (i.e. if `grp->enabled` isn't set). I believe that _should_ prevent those messages from being issued.

Does it happen all the time? Do those messages get issued every time you go through the manual recovery process? Because as I said above, they're not issued when I do it. \<shrug\>

Tell you what: it they do occur each time, try this minor tweak:

```C
            /* Prepare to wait for additional packets or pipe signal */
            FD_ZERO( &readset );
            FD_SET( grp->ppfd[0], &readset );
            if (grp->enabled)
            {
                FD_SET( grp->ttfd,    &readset );
                fd = max( grp->ppfd[0], grp->ttfd );
            }
            else
                fd = grp->ppfd[0];
```

Hopefully that'll make them go away? Please let me know if they do, and I'll make that part of my fix too.

Thanks!

----
(*) I didn't want to have to re-enable it again if I decided to run another test. All that disabling, enabling, disabling, enabling during my testing was becoming a PITA!

**Comment by Juergen-Git** [comment]
Hi Fish

 

Sorry for replying late, had a lot of other distractions to handle. Thanks for the tweak, however:

 

Yes, I disabled LSO always before attaching the new QETH device, and yes, the HHC90000D messages occur always, and no, the tweak doesn’t help because once applied it completely prevents the QETH device from coming up:

 

13:19:54 HHC03800I 0:0401 QETH: Adapter mode set to Layer 3

13:19:54 HHC04100I TunTap64.dll version 3.7.2.6156 initiated

13:19:54 HHC00137E Error opening TUN/TAP device : Invalid argument

13:19:54 HHC03996E 0:0401 QETH: : TUNTAP_CreateInterface() failed, rc=22 (0x00000016): Invalid argument

13:19:54 HHC03996E 0:0401 QETH: : qeth_create_interface() failed, rc=-1 (0xFFFFFFFF): An unidentified error has occurred

13:19:55 HHC00140E Invalid net device name

13:19:55 HHC03996E 0:0401 QETH: : qeth_enable_interface() failed, rc=0 (0x00000000): No error

13:19:55 HHC03806E 0:0401 QETH: : Cannot register guest IP address 192.168.50.246

13:19:55 HHC00905I 0:0402 QETH: Halt or clear recognized for data device

13:19:55 HHC00905I 0:0402 QETH: Halt or clear completed for data device

13:21:41 HHC00905I 0:0400 QETH: Halt or clear recognized for read device

13:21:41 HHC00905I 0:0400 QETH: Halt or clear completed for read device

 

Cheers

Jürgen

----------------------------------------------------------------------

Jürgen Winkelmann           e-mail:  ***@***.***> ***@***.***

Metallstr. 2

CH-6300 Zug

----------------------------------------------------------------------

**Comment by Fish-Git** [comment]
> Sorry for replying late, had a lot of other distractions to handle. Thanks for the tweak, however:
>
> Yes, I disabled LSO always before attaching the new QETH device, and yes, the HHC90000D messages occur always, and no, the tweak doesn’t help because once applied it completely prevents the QETH device from coming up:

(Oops!) DAMMIT! &nbsp;I don't know how it happened, but I somehow completely missed your message!

And I've already committed the changes too! &nbsp;`:(`

I'll revert the change ASAP and do some more testing. I think part of the problem on testing at MY end is that I think my network adapter's device driver may be bad. It's reporting my LSO settings incorrectly/illogically. It works perfectly on my Windows 7 virtual machine, but on my host it doesn't. That's why I suspect my host has a bad device driver.

(grumble) (this issue is starting to piss me off!) &nbsp;`:(`

Sorry folks.

**Comment by Fish-Git** [comment]
@Juergen-Git 

Which version of Visual Studio are you using to build your Hercules with? &nbsp;VS2022?

I want to be sure my testing environment is as close to yours as possible because I'm not seeing those `HHC90000D DBG: ** Win32 porting error: invalid call to 'w32_select' from qeth.c(861): mixed set(s)` messages when I perform manual recovery.

So I'm thinking maybe the version of Visual Studio that Hercules is being built with might be the reason I'm not seeing them but you are? &nbsp;_(Yes, it's a longshot, but I can't think of any other explaination!)_

**Comment by Fish-Git** [comment]
### FYI:

> (Oops!) DAMMIT!  I don't know how it happened, but I somehow completely missed your message!

I checked my gmail account very carefully and could not find _any_ email from you or github sent Feb 8th, 9th or 10th. Somebody dropped the ball, but it wasn't me!

**Comment by Fish-Git** [comment]
> `16:08:25 HHC01417I Built with: Microsoft Visual Studio 2022 (MSVC 194234433 0)`

That's the same version I'm using.


> 16:11:56 HHC04100I TunTap64.dll version 3.7.2.6156 initiated

But _that_ is _**not!**_ &nbsp;`:(`

I've been testing with my latest most current release: 3.9.0.

Let me send you privately the download URL for the latest and greatest versions of all of my products. You don't need to install all of them of course. You only need to install the latest version of CTCI-WIN.

While I'm sure I _could_ try reverting my system to the same version you're currently using, it would be easier for you to simply upgrade your install to the latest version. If that resolves it the problem, _THEN_ I can maybe research why.

**QUESTION:** does this problem prevent manual recovery from succeeding?

**Comment by Fish-Git** [comment]
> <pre>16:51:59 HHC01603I attach 0400.3 QETH <b>hwaddr 02-00-5E-A8-4B-6F</b> ipaddr 192.168.50.245 netmask 255.255.255.0 chpid F0</pre>

OR... maybe it's your `attach` statement? _(which I presume is identical to what is in your config file, yes?)_

I notice you're specifying a `hwaddr` option _(which isn't needed; why did you specify it?)_, and, oddly, are _not_ specifying the `iface` option (which is _supposed_ to be a required option on Windows!). &nbsp;I'm actually surprised it wasn't rejected as invalid!

_UNLESS..._ maybe you are specifying which Windows adapter to use via a `NETDEV` config file statement?? That must be it!

Maybe I need to take a look at both your Herc. config file as well as your `TTTest` report too. And just to be thorough, your _entire_ Herc log file too.

I still believe your problem is somehow being caused by timing. Your system is probably faster than mine, since I notice the "mixed set(s)" messages occur immediately after your `attach` command. Maybe there's a race condition somewhere in Hercules during attach processing?? I'm SO confused! &nbsp;`:(`

Are you SURE you waited until z/OS reported them ALL fully "offline" before doing the detach and attach?

Let me give your attach command syntax a try...

_(I hate it when I'm unable to reproduce a user's problem!)_

**Comment by Juergen-Git** [comment]
Hi Fish

yes, I am SURE that everything was offline before doing the detach and attach. I'll attach my config file here too. Note, that I'm using the same file for different situations, i.e. I avoid to have Windows specific definitions in the file.

Today I'm for a day out in Como (Italy), so I'll try your new CTCI-WIN version tomorrow.

Cheers
Jürgen

[zOS_3.1.txt](https://github.com/user-attachments/files/18797168/zOS_3.1.txt)

**Comment by Juergen-Git** [comment]
I've now installed CTCI-WIN 3.9.0, TTTest Report and Hercules log are attached-

[tttest64.txt](https://github.com/user-attachments/files/18814145/tttest64.txt)

[zOS_3.1.log](https://github.com/user-attachments/files/18814148/zOS_3.1.log)

I'm seeing two differences to my runs with the previous CTCI-WIN version:

- The TTTest ping test doesn't work on the first try. But from the second try onward it works flawlessly. I've set a high timeout value intentionally to make sure that the anomaly isn't caused by too short timeouts -- which it apparently isn't.

- The HHC90000D message now occurs only once, not twice as it did with the previous release.

**Comment by Fish-Git** [comment]
Hopefully fixed by commit 5f1322d4160978a1688911e59461499ea6de5093.

Please try again.

**Comment by Juergen-Git** [comment]
Hi Fish

thanks a lot for the fix. This one did the trick! No more HHC90000D messages and recovery works as expected 👍😊

Cheers
Jürgen
